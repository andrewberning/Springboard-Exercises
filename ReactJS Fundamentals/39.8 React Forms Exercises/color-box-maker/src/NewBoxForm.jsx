import React, {useState} from 'react'
import { v4 as uuid } from 'uuid';
import './NewBoxForm.css'

function NewBoxForm({ addBox }) {
  const [formData, setFormData] = useState({
    width: "",
    height: "",
    color: "",
  });

  const handleChange = (evt) => {
    const {name, value} = evt.target;
    setFormData(formData => ({
      ...formData,
      [name]: value
    }));
  };

  const handleSubmit = (evt) => {
    evt.preventDefault();
    addBox({ ...formData, id: uuid() })
    setFormData({width: "", height: "", color: "",});
  };

  return (
    <div>
      <h3>NewBoxForm Component</h3>
      <form onSubmit={handleSubmit}>
        <div>
          <label htmlFor='width'>Width: </label>
          <input 
          id='width' 
          type="text" 
          name='width'
          value={formData.width} 
          onChange={handleChange} 
          />
          <label htmlFor='height'>Height: </label>
          <input 
          id='height' 
          type="text" 
          name='height'
          value={formData.height}
          onChange={handleChange}
          />
          <label htmlFor='color'>Background Color: </label>
          <input 
          id='color' 
          type="text" 
          name='color'
          value={formData.color}
          onChange={handleChange}
          />
        </div>
        <button>Add Color Box</button>
      </form>
    </div>
  );
}

export default NewBoxForm