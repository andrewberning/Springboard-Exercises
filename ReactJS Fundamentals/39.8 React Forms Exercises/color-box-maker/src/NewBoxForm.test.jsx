import { expect, it, vi } from 'vitest';
import { render, fireEvent, waitFor } from "@testing-library/react";
import '@testing-library/jest-dom';
import NewBoxForm from './NewBoxForm';

it("renders without crashing", function () {
  render(<NewBoxForm />);
});

it("matches snapshot", function() {
  const {asFragment} = render(<NewBoxForm />);
  expect(asFragment()).toMatchSnapshot();
});

it("updates form input values when typing", function() {
  const { getByLabelText } = render(<NewBoxForm />);
  const widthInput = getByLabelText("Width:");
  const heightInput = getByLabelText("Height:");
  const colorInput = getByLabelText("Background Color:");

  fireEvent.change(widthInput, { target: { value: 5 } });
  fireEvent.change(heightInput, { target: { value: 5 } });
  fireEvent.change(colorInput, { target: { value: "red" } });

  expect(widthInput.value).toBe("5");
  expect(heightInput.value).toBe("5");
  expect(colorInput.value).toBe("red");
});

it("submits form with correct data", async function() {
  const addBoxMock = vi.fn();
  const { queryByText, getByLabelText } = render(<NewBoxForm addBox={addBoxMock} />);
  const btn = queryByText("Add Color Box");

  fireEvent.change(getByLabelText('Width:'), { target: { value: '5' } });
  fireEvent.change(getByLabelText('Height:'), { target: { value: '5' } });
  fireEvent.change(getByLabelText('Background Color:'), { target: { value: 'red' } });

  fireEvent.click(btn);

  await waitFor(() => {
    expect(addBoxMock).toHaveBeenCalledWith({ width: "5", height: "5", color: "red", id: expect.any(String)});
  });
});

it("resets form after submission", async function() {
  const addBoxMock = vi.fn();
  const { queryByText, getByLabelText } = render(<NewBoxForm addBox={addBoxMock}/>);
  const widthInput = getByLabelText("Width:");
  const heightInput = getByLabelText("Height:");
  const colorInput = getByLabelText("Background Color:");
  const btn = queryByText("Add Color Box");

  fireEvent.change(widthInput, { target: { value: 5 } });
  fireEvent.change(heightInput, { target: { value: 5 } });
  fireEvent.change(colorInput, { target: { value: "red" } });

  fireEvent.click(btn);

  await waitFor(() => {
    expect(widthInput.value).toBe("");
    expect(heightInput.value).toBe("");
    expect(colorInput.value).toBe("");
  });
});