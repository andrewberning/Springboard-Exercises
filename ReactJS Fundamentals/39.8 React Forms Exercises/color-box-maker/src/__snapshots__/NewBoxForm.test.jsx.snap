// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <h3>
      NewBoxForm Component
    </h3>
    <form>
      <div>
        <label
          for="width"
        >
          Width: 
        </label>
        <input
          id="width"
          name="width"
          type="text"
          value=""
        />
        <label
          for="height"
        >
          Height: 
        </label>
        <input
          id="height"
          name="height"
          type="text"
          value=""
        />
        <label
          for="color"
        >
          Background Color: 
        </label>
        <input
          id="color"
          name="color"
          type="text"
          value=""
        />
      </div>
      <button>
        Add Color Box
      </button>
    </form>
  </div>
</DocumentFragment>
`;
